src/main.rs

```diff
@@ -27,9 +30,44 @@ fn main() {
     let block_size = 1024;
 
     // Read audio data and write it to the output text file (one column per channel)
-    let mut out = File::create(&args[2]).expect("Unable to create file");
+    //let mut out = File::create(&args[2]).expect("Unable to create file");
+    let mut out = hound::WavWriter::create(&args[2], spec).unwrap();
+    let mut sample_index = 0;
+    let mut block_index: u32 = 0;
+    let mut input_buffer = vec![vec![f32::default(); channels as usize]; block_size as usize];
+    let mut output_buffer = vec![vec![f32::default(); channels as usize]; block_size as usize];
+    let mut comb_filter = CombFilter::new(FilterType::FIR, 1.0, 44100.0, channels as usize);
     for (i, sample) in reader.samples::<i16>().enumerate() {
         let sample = sample.unwrap() as f32 / (1 << 15) as f32;
-        write!(out, "{}{}", sample, if i % channels as usize == (channels - 1).into() { "\n" } else { " " }).unwrap();
+        let channel_index = i % channels as usize;
+        if sample_index < block_size {
+            input_buffer[sample_index][channel_index] = sample;
+        } else {
+            let input_buffer_slice = &input_buffer[..].iter().map(|v| v.as_slice()).collect::<Vec<_>>();
+            let input_buffer_slice_2d = &input_buffer_slice[..];
+            let mut output_buffer_slice = output_buffer[..].iter_mut().map(|v| v.as_mut_slice()).collect::<Vec<_>>();
+            let output_buffer_slice_2d = &mut output_buffer_slice[..];
+            comb_filter.process(input_buffer_slice_2d, output_buffer_slice_2d);
+            for block_index in output_buffer_slice_2d.iter_mut() {
+                for channel in block_index.iter_mut() {
+                    let written_sample = (*channel * i16::MAX as f32) as i16;
+                    out.write_sample(written_sample).unwrap();
+                }
+            }
+            sample_index = 0;
+            input_buffer[sample_index][channel_index] = sample;
+            block_index += 1;
+        }
+        if i % channels as usize == (channels - 1) as usize {
+            sample_index += 1;
+        }
+    }
+    let current_length = (block_index*block_size as u32);
+    let remaining_length = (reader.len()/channels as u32) - current_length;
+    let mut i: u32 = 0;
+    while i < remaining_length {
+        out.write_sample(0).unwrap();
+        out.write_sample(0).unwrap();
+        i += 1;
     }
 }
```

Comment:
I am a bit unsure but I wanted to share related to your statement about saving the output of the blocks.  
// If you use your code to process the blocks with comb_filter....
// Process input block with FIR comb filter
       
 comb_filter.process(input_buffer_slice_2d, output_buffer_slice_2d);

// You should be able to do something like this
// Write output block to the output text file
        
for (i, sample) in output_buffer.iter().enumerate() {
            write!(out, "{}{}", sample, if i % channels == channels - 1 { "\n" } else { " " }).unwrap();

The write out is similar to what we did in exercise 2. 

I used some of my code so you may need to revise based on your variable. I hope this is helpful.

Reply:
I decided to check if there are remaining samples and then process another block with these remaining samples zero padded.